File: src/parsing/expressions/binding_reference.ts
import Bindings from "../../graph/bindings";
import Identifier from "./identifier";

/**
 * Reference to a `LET`-bound binding by name.
 *
 * Distinct from {@link Reference} (which points to a scope-local
 * variable in the parser state) — a `BindingReference` is resolved
 * against the global {@link Bindings} singleton at execution time.
 */
class BindingReference extends Identifier {
    /**
     * When `true`, resolving a name that is not present in the global
     * {@link Bindings} store yields `undefined` instead of throwing.
     *
     * Used for projection references (`RETURN name` / `WITH name`),
     * where a bare unresolved identifier historically evaluated to
     * `undefined`; lenient mode preserves that behaviour for genuinely
     * unknown names while still resolving real `LET` bindings.
     */
    private _optional: boolean;

    constructor(value: string, optional: boolean = false) {
        super(value);
        this._optional = optional;
    }

    public get name(): string {
        return this._value;
    }

    public toString(): string {
        return `BindingReference (${this._value})`;
    }

    public value(): any {
        const bindings = Bindings.getInstance();
        if (!bindings.has(this._value)) {
            if (this._optional) {
                return undefined;
            }
            throw new Error(`Binding '${this._value}' is not defined`);
        }
        return bindings.get(this._value);
    }
}

export default BindingReference;

File: flowquery-py/src/parsing/expressions/binding_reference.py
"""Reference to a LET-bound binding by name."""

from typing import Any

from .identifier import Identifier


class BindingReference(Identifier):
    """Reference to a LET-bound binding, resolved at execution time
    against the global :class:`Bindings` singleton."""

    def __init__(self, value: str, optional: bool = False):
        super().__init__(value)
        # When True, resolving a name absent from the global Bindings
        # store yields None instead of raising.  Used for projection
        # references (``RETURN name`` / ``WITH name``), where a bare
        # unresolved identifier historically evaluated to None; lenient
        # mode preserves that behaviour for genuinely unknown names
        # while still resolving real LET bindings.
        self._optional = optional

    @property
    def name(self) -> str:
        return self._value

    def __str__(self) -> str:
        return f"BindingReference ({self._value})"

    def value(self) -> Any:
        from ...graph.bindings import Bindings

        bindings = Bindings.get_instance()
        if not bindings.has(self._value):
            if self._optional:
                return None
            raise RuntimeError(f"Binding '{self._value}' is not defined")
        return bindings.get(self._value)
